pyteal/ast/abi/tuple.py

```diff
@@ -484,3 +486,57 @@ def __init__(
 
 
 Tuple5.__module__ = "pyteal.abi"
+
+
+class NamedTupleTypeSpec(TupleTypeSpec):
+    """A NamedTupleType inherits from TupleTypeSpec, allowing for more than 5 elements."""
+
+    def __init__(
+        self, instance_class: type["NamedTuple"], *value_type_specs: TypeSpec
+    ) -> None:
+        if instance_class == NamedTuple:
+            raise TealInputError(
+                "NamedTupleTypeSpec must be instanced with subclassed NamedTuple class."
+            )
+
+        self.instance_class: type["NamedTuple"] = instance_class
+        super().__init__(*value_type_specs)
+
+    def annotation_type(self) -> "type[NamedTuple]":
+        return self.instance_class
+
+    def new_instance(self) -> "NamedTuple":
+        return self.instance_class()
+
+
+NamedTupleTypeSpec.__module__ = "pyteal.abi"
+
+
+class NamedTuple(Tuple):
+    """A NamedTuple is a Tuple with all its elements named."""
+
+    def __init__(self):
+        if type(self) is NamedTuple:
+            raise TealInputError("NamedTuple must be subclassed.")
+
+        anns = get_annotations(type(self))
+        if not anns:
+            raise Exception("Expected fields to be declared but found none")
+
+        self.type_specs: OrderedDict[str, TypeSpec] = OrderedDict()
+
+        for index, (name, annotation) in enumerate(anns.items()):
+            self.type_specs[name] = type_spec_from_annotation(annotation)
+
+        super().__init__(
+            NamedTupleTypeSpec(type(self), *list(self.type_specs.values()))
+        )
+
+        for index, name in enumerate(anns):
+            setattr(self, name, self[index])
+
+    def type_spec(self) -> TupleTypeSpec:
+        return cast(NamedTupleTypeSpec, super().type_spec())
+
+
+NamedTuple.__module__ = "pyteal.abi"
```

Comment:
I'd like to see a unit test in `tuple_test.py` that confirms the return value of each user-defined field is as expected, something like the existing `test_Tuple_getitem` test

Reply:
good point, just add a similar test based on NamedTuple